### ISSUE TITLE
Why are the attributes marked as internal?

### ISSUE THREAD

**Issue by mrdnote** [manuelroemer/Nullable#2]
I've added the project to my solution, but I can't use the attributes since they are marked as internal. When I change them to public it works brilliantly.

**Comment by manuelroemer** [comment]
The attributes are internal, so that no name conflicts can happen. 
If the attributes were `public`, naming conflicts should happen. For example, have a look at this (very professional looking) constellation:
![grafik](https://user-images.githubusercontent.com/30902964/65955203-e78b7c80-e447-11e9-8caf-22ea37c23b12.png)

The library at the top uses the package. If the attribute is `public` it becomes visible to other projects which use this library. A problem occurs when there are two attributes with the same name, e.g. for a .NET Core 3.0 app. Now there are two attributes in the same namespace with the same name - one from .NET Core and one from the library. The compiler will be unable to choose the correct one and thus won't compile.

By making the attribute `internal`, it is not visible to others, hence no conflict can occur (in this example, the .NET Core app would directly use the attribute defined by the .NET Core BCL).

I am curious why your project doesn't work if the attributes are `internal` though. Could you share your exact project setup, so that I can have a look at it? Or can you tell me how I can reproduce the issue?

Sorry to hear that you seem to have such a rough time getting this to work - let's solve these problems!

**Comment by manuelroemer** [comment]
By the way, I am thinking of introducing a compiler constant like `NULLABLE_ATTRIBUTES_PUBLIC` which change the modifiers from `internal` to `public`. This is something that [LibLog](https://github.com/damianh/LibLog) does (LibLog also adds source code files to projects).
This would 100% be opt-in though, because `internal` should always be the default.

**Comment by mrdnote** [comment]
Classic .NET framework projects don't have conflicts, because the System.Diagnostics.CodeAnalysis namespace does not exist there.

And this is also why I can't use the attributes, because I have no access to System.Diagnostics.CodeAnalysis. The internal classes cannot be specified because they are internal to their own project.

In a .NET Core app situation, how come your attributes are being used while the are marked as inaccessible, so in effect the code is referencing the .NET Core framework attributes instead of your Nullable package's attributes? Or are they just there to "fool" the compiler?

**Comment by manuelroemer** [comment]
Even classic .NET Framework projects could in theory get these conflicts. Here is an example:

```
LibraryA:
  - public class AllowNullAttribute

LibraryB:
  - public class AllowNullAttribute

App:
  * Uses LibraryA
  * Uses LibraryB
```

If you then wrote this code in `App`, you would also get a compiler error, because the compiler doesn't know if he should use the attribute from `LibraryA` or `LibraryB`:

```csharp
using System.Diagnostics.CodeAnalysis;

[AllowNull] public string Foo { get; set; }
```

This is why the attributes are defined as internal - if they don't "leak out" of the project, there will not be any conflict.

> Or are they just there to "fool" the compiler?

That's basically it. The whole idea works, because the C# compiler simply checks if there is an attribute with the same **name** and **namespace** when doing nullable analysis, but the attribute does not have to come from the official .NET BCL. Because of this, it even works if the attribute is internal. And if you use this package with a .NET Core 3.0 app, the `Nullable.cs` file simply doesn't get added to the project, because .NET already defines these attributes for you.

Ideally, projects using this package are setup like this:

```
LibraryA:
  * References Nullable NuGet package
  - internal class AllowNullAttribute

LibraryB: 
  * References Nullable NuGet package
  - internal class AllowNullAttribute

App:
  * Uses LibraryA
  * Uses LibraryB
  * References Nullable NuGet package
  - internal class AllowNullAttribute
```

This way, the attributes exist in _every project_, but don't leak out. The compiler can still do nullable analysis, because each project still defines the attributes internally.

To get back to the original problem: Are your projects set-up like above, i.e. does each project reference the NuGet package separately? If not, how did you add the attributes?

**Comment by mrdnote** [comment]
I know how conflicts work, I'm just saying that I don't have them in this case, because the AllowNullAttribute is not defined for .NET Framework. Your definition is the first and only one from the perspective of my project.

My project is setup so each library and app references your Nullable project.

Attached you can find a test project which demonstrates the issue.
[NullableTestConsoleApp.zip](https://github.com/manuelroemer/Nullable/files/3675906/NullableTestConsoleApp.zip)

**Comment by manuelroemer** [comment]
Thanks, that project is very helpful! 👍 

The reason for why it doesn't work is that the Nullable project has not been added via NuGet. Try the following steps for a testing project:

1. Create a new .NET Console project using the old .NET FW.
2. Go into the Visual Studio Options > NuGet Package Manager and select the "Allow format selection on first package install" checkbox. This allows you to choose if you want to use a packages.config file or package references when you install your first package.  
  ![grafik](https://user-images.githubusercontent.com/30902964/65958953-f9255200-e450-11e9-8705-f8cd899bc74e.png)

3. Go into the package manager by right clicking "References" > "Manage NuGet packages...". Search for "Nullable" and install the package. In the dialog which pops up, choose the "Package reference in project file" option.  
  ![grafik](https://user-images.githubusercontent.com/30902964/65958978-06424100-e451-11e9-80bf-067360480a7f.png)
4. Enable C# 8.0 so that you have Nullable support by unloading and editing your project and then editing the XML like in the third image:
  ![grafik](https://user-images.githubusercontent.com/30902964/65959014-2540d300-e451-11e9-8294-9b7d20f08379.png)  
  ![grafik](https://user-images.githubusercontent.com/30902964/65959023-2d990e00-e451-11e9-974c-c3e4f5b0d5fc.png)  
  ![grafik](https://user-images.githubusercontent.com/30902964/65959041-3e498400-e451-11e9-8fdd-c8c7049dd98f.png)
5. Reload your project - you should now be able to use the attributes in existing code.  
  ![grafik](https://user-images.githubusercontent.com/30902964/65959054-443f6500-e451-11e9-8bbe-656123cc75a6.png)
6. Try it out by, for example, pasting the code below and checking if the warnings are shown:
  ![grafik](https://user-images.githubusercontent.com/30902964/65959122-76e95d80-e451-11e9-92f7-2fd73bf1d855.png)

Furthermore, you can put the cursor on the attribute and press `F12` - this way, you can see if the file actually got added to your project:

Click `F12` here:
![grafik](https://user-images.githubusercontent.com/30902964/65959409-32aa8d00-e452-11e9-9311-804a19a73380.png)

and you will see this:
![grafik](https://user-images.githubusercontent.com/30902964/65959433-40601280-e452-11e9-8198-3f2217886640.png)

You can download and test the finished demo project here:
[NullableNetFw.zip](https://github.com/manuelroemer/Nullable/files/3676022/NullableNetFw.zip)

I hope that this helps (and works) for you!

**Comment by mrdnote** [comment]
Yeah, in my other https://github.com/manuelroemer/Nullable/issues/1 I already gave away that I would add Nullable.csproj as a project to my solution. Now I understand the miscommunication about internal/public and nuget package references :D

**Comment by manuelroemer** [comment]
Okay awesome, I'm glad that everything is settled now! :)

> I don't expect your project to be high maintenance with a lot of updates.

Yeah, I agree - there are not going to be many changes here. 

One last tip: You can consider to just create a [Shared Project](https://docs.microsoft.com/en-us/xamarin/cross-platform/app-fundamentals/shared-projects?tabs=windows#creating-a-shared-project-1) in Visual Studio and then copy only the [NullableAttributes.cs](https://github.com/manuelroemer/Nullable/blob/master/src/Nullable/NullableAttributes.cs) file and nothing else to it. This way, you won't even have to change the `internal` to `public` and you won't have to deal with NuGet packages.
That is, in my opinion, the best out of both worlds.
